cmd/common/conn/tls.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2020 Authors of Hubble
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package conn
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/cilium/hubble/pkg/defaults"
+	"github.com/spf13/viper"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+func init() {
+	GRPCOptionFuncs = append(GRPCOptionFuncs, grpcOptionTLS)
+}
+
+func grpcOptionTLS(vp *viper.Viper) (grpc.DialOption, error) {
+	target := vp.GetString("server")
+	if !(vp.GetBool("tls") || strings.HasPrefix(target, defaults.TargetTLSPrefix)) {
+		return grpc.WithInsecure(), nil
+	}
+
+	tlsConfig := tls.Config{
+		InsecureSkipVerify: vp.GetBool("tls-allow-insecure"),
+		ServerName:         vp.GetString("tls-server-name"),
+	}
+
+	// optional custom CAs
+	caFiles := vp.GetStringSlice("tls-ca-cert-files")
+	if len(caFiles) > 0 {
+		ca := x509.NewCertPool()
+		for _, path := range caFiles {
+			certPEM, err := ioutil.ReadFile(path)
+			if err != nil {
+				return nil, fmt.Errorf("cannot load cert '%s': %s", path, err)
+			}
+			if ok := ca.AppendCertsFromPEM(certPEM); !ok {
+				return nil, fmt.Errorf("cannot process cert '%s': must be a PEM encoded certificate", path)
+			}
+		}
+		tlsConfig.RootCAs = ca
+	}
+
+	// optional mTLS
+	clientCertFile := vp.GetString("tls-client-cert-file")
+	clientKeyFile := vp.GetString("tls-client-key-file")
+	if clientCertFile != "" && clientKeyFile != "" {
+		cert, err := tls.LoadX509KeyPair(clientCertFile, clientKeyFile)
+		if err != nil {
+			return nil, err
+		}
+		tlsConfig.Certificates = []tls.Certificate{cert}
+	}
```

Comment:
IMO it would be nice to use Cilium's `pkg/crypto/certloader` stuff here so that we won't have to implement the `x509.NewCertPool()` and `tls.LoadX509KeyPair()` dance all over again.

At the moment, Cilium's certloader stuff is always running a file watcher which in the context of Hubble CLI is overkill. Maybe it would be a good idea to extract the certloader in its own go module / repository so that we can extend it with what CLI needs.

In the mean time (or in any case), this PR could port [how the certloader client is handling the keypair using the `GetClientCertificate` callback rather than the `tlsConfig.Certificates` directly](https://github.com/cilium/cilium/blob/master/pkg/crypto/certloader/client.go#L66-L69). The value here is when the server is configured for mTLS but a cli invocation fail to provide the client keypair. This is what we get with the patch as it stand:

```
% ./hubble observe \
    --server tls://localhost:4244 \
    --tls-ca-cert-files ca.pem \
    --tls-server-name foo.default.hubble-grpc.cilium.io
failed to connect to 'tls://localhost:4244': context deadline exceeded: connection closed
```

And this is what we could get using the `GetClientCertificate` callback:

```
% ./hubble observe \
    --server tls://localhost:4244 \
    --tls-ca-cert-files ca.pem \
    --tls-server-name foo.default.hubble-grpc.cilium.io
failed to connect to 'tls://localhost:4244': context deadline exceeded: connection error: desc = "transport: authentication handshake failed: mTLS client certificate requested, but not configured"
```